lib/resty/core/socket_tcp.lua

```diff
@@ -0,0 +1,195 @@
+    local client_priv_key = options.client_priv_key
+    if client_cert then
+        if not client_priv_key then
+            error("client certificate supplied without "
+                  .. "corresponding private key", 2)
+        end
+
+        if type(client_cert) ~= "cdata"
+           or type(client_priv_key) ~= "cdata"
+        then
+            error("wrong type of client certificate or private key supplied", 2)
+        end
+    end
+
+    local rc =
+        C.ngx_http_lua_ffi_socket_tcp_tlshandshake(r, self[SOCKET_CTX_INDEX],
+                                                   session_ptr[0],
+                                                   reused_session ~= false,
+                                                   server_name_str,
+                                                   options.verify and 1 or 0,
+                                                   options.ocsp_status_req
+                                                       and 1 or 0,
+                                                   client_cert,
+                                                   client_priv_key,
+                                                   errmsg)
+
+    if rc == FFI_NO_REQ_CTX then
+        error("no request ctx found", 2)
+    end
+
+::again::
+
+    if rc == FFI_ERROR then
+        if openssl_error_code[0] ~= 0 then
+            return nil, openssl_error_code[0] .. ": " .. ffi_string(errmsg[0])
+        end
+
+        return nil, ffi_string(errmsg[0])
+    end
+
+    if rc == FFI_DONE then
+        return options.reused_session
+    end
+
+    if rc == FFI_OK then
+        if options.reused_session == false then
+            return true
+        end
+
+        rc = C.ngx_http_lua_ffi_socket_tcp_get_tlshandshake_result(r,
+            self[SOCKET_CTX_INDEX], session_ptr, errmsg, openssl_error_code)
+
+        assert(rc == FFI_OK)
+
+        if session_ptr[0] == nil then
+            return session_ptr[0]
+        end
+
+        return ffi_gc(session_ptr[0], C.ngx_http_lua_ffi_tls_free_session)
+    end
+
+    assert(rc == FFI_AGAIN)
+
+    co_yield()
+
+    rc = C.ngx_http_lua_ffi_socket_tcp_get_tlshandshake_result(r,
+        self[SOCKET_CTX_INDEX], session_ptr, errmsg, openssl_error_code)
+
+    assert(rc == FFI_OK or rc  == FFI_ERROR)
+
+    goto again
+end
+
+
+local function sslhandshake(self, reused_session, server_name, ssl_verify,
+    send_status_req, ...)
+
+    local n = select("#", ...)
+    if not self or n > 1 then
+        error("ngx.socket sslhandshake: expecting 1 ~ 5 "
+              .. "arguments (including the object), but seen " .. n)
+    end
+
+    cached_options.reused_session = reused_session
+    cached_options.server_name = server_name
+    cached_options.verify = ssl_verify
+    cached_options.ocsp_status_req = send_status_req
+
+    local res, err = tlshandshake(self, cached_options)
+    table_clear(cached_options)
+
+    return res, err
+end
+
+
+do
+    local old_socket_tcp = ngx.socket.tcp
```

Comment:
I have three comments about this monkeypatching.

* It doesn't seem to be done in other parts of `lua-resty-core`.
* It obscures the origin of the `ngx.socket.tcp` function . This might trip someone who (reasonably) expects ngx_http_lua_socket_tcp.h/c to be the "source of truth".
* It is not a very solid base on which to build upon. Would it be acceptable to have 4 different lua files, each one monkeypatching `ngx.socket.tcp` in a different way?

My suggestion would be to move `tlshandshake` and `sslhandshake` to the sister PR, https://github.com/openresty/lua-nginx-module/pull/1602 (perhaps translating them to C?) and adding them to the sockets metatable there.

Alternatively, I would expose the two Lua functions from here, and then I would include them inside the metatable in https://github.com/openresty/lua-nginx-module/pull/1602. I'm unclear if that's even possible.

Reply:
Adding more Lua C functions is exactly what we do not want to do. The goal is to implement new cosocket methods using FFI without having to reimplement all the existing ones at the same time (because the PR will be unacceptably large and time consuming).

Eventually this will go away, after we migrated all the methods to use FFI instead. 

Also see: https://github.com/openresty/lua-nginx-module/pull/1599#discussion_r325342424

Reply:
Ok. Forget about implementing them in C then.

Would it be possible to implement them in FFI but *without* having to do this monkeypatching? The reason why it's necessary seems to be to be getting a hold of the socket's metatable. Perhaps it could be exposed directly, somewhere? `ngx.socket.tcp_mt`. (Yeah I agree that is a bit ugh. Maybe there is a better place).

But at the same time, exposing the metatable is only necessary in order to add methods to sockets. We could add a "hook" which does just that. Something like:

```
ngx.socket.tcp.new_method("tlshandshake", tlshandshake)
```

I like that this is very greppable and explicit. However it requires transforming into a "callable table" on the sister PR. If that is too much complication, this is my second-best suggestion:

```
ngx.socket.new_method("tcp", "tlshandshake", tlshandshake)
```